oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/web/server/WebSessionOAuth2ServerAuthorizationRequestRepositoryTests.java

```diff
@@ -99,8 +103,175 @@ public void loadAuthorizationRequestWhenSavedThenAuthorizationRequest() {
+		}
+
+		@Override
+		public Map<String, Object> getAttributes() {
+			return attributes;
+		}
+
+		@Override
+		public Mono<Void> save() {
+			try {
+				if (attributes.isDirty()) {
+					attributes.resolve();
+					attributes.reset();
+					return super.save();
+				} else {
+					attributes.reset();
+				}
+			} catch (InstantiationException | IllegalAccessException ignore) {
+			}
+
+			return Mono.empty();
+		}
+	}
+
+	/**
+	 * A simulation to capture the session data change.
+	 */
+	private static class DeltaOptimizedMap implements Map<String, Object> {
+		private Map<String, Object> delegated;
+		private HashMap<String, Object> cached = new HashMap<>();
+		boolean dirty = false;
+
+		public DeltaOptimizedMap(Map<String, Object> delegated)
+				throws InstantiationException, IllegalAccessException {
+			this.delegated = delegated;
+			cloneMap(delegated, cached);
+		}
+
+		public void resolve() {
+			if (dirty) {
+				delegated.clear();
+				delegated.putAll(cached);
+			}
+		}
+
+		// reset session data as if creating a new session object with the same session id.
+		public void reset() throws InstantiationException, IllegalAccessException {
+			dirty = false;
+			cloneMap(delegated, cached);
+		}
+
+		private void cloneMap(Map<String, Object> src, Map<String, Object> dest)
+				throws IllegalAccessException, InstantiationException {
+			dest.clear();
+
+			for (String key: src.keySet()){
+				Object value = src.get(key);
+
+				if (value instanceof Map) {
+					Map newValue = (Map) value.getClass().newInstance();
+					cloneMap((Map<String, Object>) value, newValue);
+					dest.put(key, newValue);
+				} else {
+					dest.put(key, value);
+				}
+			}
+		}
+
+		public boolean isDirty() {
+			return dirty;
+		}
+
+		@Override
+		public int size() {
+			return cached.size();
+		}
+
+		@Override
+		public boolean isEmpty() {
+			return cached.isEmpty();
+		}
+
+		@Override
+		public boolean containsKey(Object key) {
+			return cached.containsKey(key);
+		}
+
+		@Override
+		public boolean containsValue(Object value) {
+			return cached.containsValue(value);
+		}
+
+		@Override
+		public Object get(Object key) {
+			return cached.get(key);
+		}
+
+		@Override
+		public Object put(String key, Object value) {
+			dirty = true;
+			return cached.put(key, value);
+		}
+
+		@Override
+		public Object remove(Object key) {
+			dirty = true;
+			return cached.remove(key);
+		}
+
+		@Override
+		public void putAll(Map<? extends String, ?> m) {
+			dirty = true;
+			cached.putAll(m);
+		}
+
+		@Override
+		public void clear() {
+			dirty = true;
+			cached.clear();
+		}
+
+		@Override
+		public Set<String> keySet() {
+			return cached.keySet();
+		}
+
+		@Override
+		public Collection<Object> values() {
+			return cached.values();
+		}
+
+		@Override
+		public Set<Entry<String, Object>> entrySet() {
+			return cached.entrySet();
+		}
+	}
+
 	@Test
-	public void loadAuthorizationRequestWhenMultipleSavedThenAuthorizationRequest() {
```

Comment:
Please do not modifiy any of the existing tests as this means that we might have broken something. Please instead add a new test that does not change existing tests